ouroboros-consensus-protocol/src/ouroboros-consensus-protocol/Ouroboros/Consensus/Protocol/Praos/Common.hs

```diff
@@ -244,16 +255,57 @@ instance Crypto c => ChainOrder (PraosChainSelectView c) where
   preferCandidate cfg ours cand = comparePraos cfg ours cand == LT
 
 data PraosCanBeLeader c = PraosCanBeLeader
-  { -- | Certificate delegating rights from the stake pool cold key (or
-    -- genesis stakeholder delegate cold key) to the online KES key.
-    praosCanBeLeaderOpCert     :: !(OCert.OCert c),
-    -- | Stake pool cold key or genesis stakeholder delegate cold key.
-    praosCanBeLeaderColdVerKey :: !(SL.VKey 'SL.BlockIssuer),
-    praosCanBeLeaderSignKeyVRF :: !(VRF.SignKeyVRF (VRF c))
+  { -- | Stake pool cold key or genesis stakeholder delegate cold key.
+    praosCanBeLeaderColdVerKey        :: !(SL.VKey 'SL.BlockIssuer),
+    praosCanBeLeaderSignKeyVRF        :: !(SignKeyVRF (VRF c)),
+    -- | How to obtain KES credentials (ocert + sign key)
+    praosCanBeLeaderCredentialsSource :: !(PraosCredentialsSource c)
   }
   deriving (Generic)
 
-instance Crypto c => NoThunks (PraosCanBeLeader c)
+instance (NoThunks (SignKeyVRF (VRF c)), NoThunks (KES.UnsoundPureSignKeyKES (KES c)), Crypto c) => NoThunks (PraosCanBeLeader c)
+
+-- | Defines a method for obtaining Praos credentials (opcert + KES signing
+-- key).
+data PraosCredentialsSource c
+  = -- | Pass an opcert and sign key directly. This uses
+    -- 'KES.UnsoundPureSignKeyKES', which does not provide mlocking guarantees,
+    -- violating the rule that KES secrets must never be stored on disk, but
+    -- allows the sign key to be loaded from a local file. This method is
+    -- provided for backwards compatibility.
+    PraosCredentialsUnsound (OCert.OCert c) (KES.UnsoundPureSignKeyKES (KES c))
+  | -- | Connect to a KES agent listening on a service socket at the given path.
+    PraosCredentialsAgent FilePath
+  deriving (Generic)
+
+instance (NoThunks (KES.UnsoundPureSignKeyKES (KES c)), Crypto c) => NoThunks (PraosCredentialsSource c)
+
+instantiatePraosCredentials :: forall m c.
+                               ( KESAgentContext c m
+                               )
+                            => Word64
+                            -> PraosCredentialsSource c
+                            -> m (HotKey.HotKey c m)
+instantiatePraosCredentials maxKESEvolutions (PraosCredentialsUnsound ocert skUnsound) = do
+  sk <- KES.unsoundPureSignKeyKESToSoundSignKeyKES skUnsound
+  let startPeriod :: OCert.KESPeriod
+      startPeriod = OCert.ocertKESPeriod ocert
+
+  HotKey.mkHotKey
+              ocert
+              sk
+              startPeriod
+              maxKESEvolutions
+
+instantiatePraosCredentials maxKESEvolutions (PraosCredentialsAgent path) = do
+  HotKey.mkDynamicHotKey
+      maxKESEvolutions
+      (Just $ \send -> do
+        let handleKey ocert sk p = do
+              send ocert sk p (OCert.ocertKESPeriod ocert)
+        runKESAgentClient nullTracer path handleKey
```

Comment:
Here we should probably pass the tracer above?

Reply:
Indeed, but I'm not entirely sure how to do this cleanly, since `instantiatePraosCredentials` is used from within the block forgings defined in `protocolInfoCardano`, which doesn't (and shouldn't) have `Tracers` in scope. The best I can come up with at this point is extend `protocolInfoCardano` such that the block forging generator it returns takes the tracer as an extra argument, but that change would then ripple to all call sites of `protocolInfoCardano`.

Reply:
If we need it, then so be it